### ISSUE TITLE
Fix for #2116 causes an exception for non-existing users

### ISSUE THREAD

**Issue by ssg** [IdentityServer/IdentityServer4#3393]
### Issue / Steps to reproduce the problem
Try to get an access token using user/password authentication for a non-existing user.

Actual result: 

```
crit: IdentityServer4.Hosting.IdentityServerMiddleware[0]
      Unhandled exception: Error
      Parameter name: Error must be set
System.ArgumentNullException: Error
Parameter name: Error must be set
   at IdentityServer4.Endpoints.Results.TokenErrorResult..ctor(TokenErrorResponse error) in C:\ballen\github\identity\IdSvr4\IdentityServer4\src\IdentityServer4\Endpoints\Results\TokenErrorResult.cs:line 20
   at IdentityServer4.Endpoints.TokenEndpoint.Error(String error, String errorDescription, Dictionary`2 custom) in C:\ballen\github\identity\IdSvr4\IdentityServer4\src\IdentityServer4\Endpoints\TokenEndpoint.cs:line 117
   at IdentityServer4.Endpoints.TokenEndpoint.ProcessTokenRequestAsync(HttpContext context) in C:\ballen\github\identity\IdSvr4\IdentityServer4\src\IdentityServer4\Endpoints\TokenEndpoint.cs:line 93
   at IdentityServer4.Endpoints.TokenEndpoint.ProcessAsync(HttpContext context) in C:\ballen\github\identity\IdSvr4\IdentityServer4\src\IdentityServer4\Endpoints\TokenEndpoint.cs:line 70
   at IdentityServer4.Hosting.IdentityServerMiddleware.Invoke(HttpContext context, IEndpointRouter router, IUserSession session, IEventService events) in C:\ballen\github\identity\IdSvr4\IdentityServer4\src\IdentityServer4\Hosting\IdentityServerMiddleware.cs:line 54
```

Expected result: 

```
{
    "error": "invalid_grant",
    "error_description": "invalid_username_or_password",
    "description": "Invalid user name or password."
}
```

Rolling back to 2.1.2 fixes this issue. Any version with 2.1.3 or above has this problem.

**Comment by brockallen** [comment]
Do you have a custom `IResourceOwnerPasswordValidator` implemented?

**Comment by ssg** [comment]
@brockallen Yes, apparently we created an empty `GrantValidationResult` class (instead of one from a `TokenRequestErrors.InvalidGrant`) and set `IsError` `true` but not the `Error` itself. Setting `Error` fixed this. However, this change to require `Error` to be set after v2.1.3 wasn't apparent from the usage. Perhaps `GrandValidationResult` shouldn't allow this invalid state (if having `IsError==true && Error==null` is an invalid state). I changed our constructor to the one with `TokenRequestErrors.InvalidGrant`, so it's not an urgent issue for our specific case anymore, thanks!.
